### PR TITLE
[CECO-743] Secrets backend feature

### PR DESCRIPTION
**Pull request by tbavelier**
### What does this PR do?

Implements the Secrets Backend feature to directly configure [Secrets management](https://docs.datadoghq.com/agent/configuration/secrets-management/?tab=linux) from the `DatadogAgent` custom resource, providing helpers to handle RBAC similar to the Helm chart

### Motivation

* https://datadoghq.atlassian.net/browse/CECO-743
* Helm chart parity
* Internal feedback

### Describe your test plan
The scenarios below are covered by the unit tests and replicate them in e2e manner, verifying functionality. Ensure the new version of the CRD is installed with `make install` and install the built-operator with `make IMG=tbdatadog/operator:1.9.0 deploy` (pre-loading your built image in your cluster with `kind load docker-image tbdatadog/operator:1.9.0 --name <KIND CLUSTER NAME>`
#### Testing env variables
```yaml
[...]
  features:
    [...]
    secretBackend:
      command: "/readsecret_multiple_providers.sh"
      args: "foo bar"
      timeout: 60
```
Exec into your Agent pod and assert:
- [ ] `agent config | grep secret_backend -A 2` matches the parameters defined in the CR
image

#### Testing global RBAC and secrets resolution
* Remove the `args` and `timeout`. Add `enableGlobalPermissions` set to `true`. Re-deploy your CR
```yaml
[...]
  features:
    [...]
    secretBackend:
      command: "/readsecret_multiple_providers.sh"
      enableGlobalPermissions: true
```
* Deploy the RabbitMQ operator and a sample workload : 
    * `kubectl apply -f "https://github.com/rabbitmq/cluster-operator/releases/download/v2.9.0/cluster-operator.yml"`
    * Apply the manifest below 
      ```yaml
      apiVersion: rabbitmq.com/v1beta1
      kind: RabbitmqCluster
      metadata:
        name: rabbitmqcluster-sample
        namespace: rabbitmq-system
      spec:
        override:
          statefulSet:
            spec:
              template:
                metadata:
                  annotations:
                    ad.datadoghq.com/rabbitmq.check_names: '["rabbitmq"]'
                    ad.datadoghq.com/rabbitmq.init_configs: "[{}]"
                    ad.datadoghq.com/rabbitmq.instances: |
                      [
                        {
                          "rabbitmq_api_url": "http://%%host%%:15672/api/",
                          "username": "ENC[k8s_secret@rabbitmq-system/rabbitmqcluster-sample-default-user/username]",
                          "password": "ENC[k8s_secret@rabbitmq-system/rabbitmqcluster-sample-default-user/password]"
                        }
                      ]
      ```
Assert the following : 
- [ ] Both the Agent and DCA are able to access the k8s secret with `k auth can-i get -n rabbitmq-system secrets/rabbitmqcluster-sample-default-user --as=system:serviceaccount:system:datadog-agent`
    image
- [ ] The Agent is resolving the secret successfully with `agent secret` inside the node Agent
    image

#### Testing specific RBAC (roles), its priority over `enableGlobalPermissions` and binding
* Ensure `rabbitmq-system` is within `WATCH_NAMESPACE` variable if not using global watch scope : 
    image
* Add roles and enable CCR. Re-deploy your CR
```yaml
[...]
  features:
    [...]
     clusterChecks:
      useClusterChecksRunners: true
    secretBackend:
      command: "/readsecret_multiple_providers.sh"
      enableGlobalPermissions: true
      roles:
      - namespace: rabbitmq-system
        secrets:
        - "rabbitmqcluster-sample-default-user"

```
Assert the following : 
- [ ] The Agent, DCA and CCR are able to access the k8s secret with `k auth can-i get -n rabbitmq-system secrets/rabbitmqcluster-sample-default-user --as=system:serviceaccount:system:datadog-agent`
- [ ] The Agent, DCA and CCR are **NOT able to access the non-specified k8s secret** with `k auth can-i get -n rabbitmq-system secrets/rabbitmqcluster-sample-erlang-cookie --as=system:serviceaccount:system:datadog-agent`
image

### Checklist

- [x] PR has at least one valid label: `bug`, `enhancement`, `refactoring`, `documentation`, `tooling`, and/or `dependencies`
- [x] PR has a milestone or the `qa/skip-qa` label

### CODE REVIEW

File: docs/configuration.v2alpha1.md [hunk 1]
```diff
@@ -156,6 +156,11 @@ spec:
 | features.sbom.enabled | Enable this option to activate SBOM collection. Default: false |
 | features.sbom.host.analyzers | Analyzers to use for SBOM collection. |
 | features.sbom.host.enabled | Enable this option to activate SBOM collection. Default: false |
+| features.secretsBackend.args | Args defines the list of arguments to pass to the command (space-separated strings). |
```
Comment (neko-dd) [inline docs/configuration.v2alpha1.md:159]:
```suggestion
| features.secretsBackend.args | List of arguments to pass to the command (space-separated strings). |
```

File: docs/configuration.v2alpha1.md [hunk 2]
```diff
@@ -156,6 +156,11 @@ spec:
...
+| features.secretsBackend.command | Command defines the secret backend command to use. Datadog provides a pre-defined binary `/readsecret_multiple_providers.sh`. Read more about `/readsecret_multiple_providers.sh`: https://docs.datadoghq.com/agent/configuration/secrets-management/?tab=linux#script-for-reading-from-multiple-secret-providers |
```
Comment (neko-dd) [inline docs/configuration.v2alpha1.md:160]:
```suggestion
| features.secretsBackend.command | The secret backend command to use. Datadog provides a pre-defined binary `/readsecret_multiple_providers.sh`. Read more about `/readsecret_multiple_providers.sh` at https://docs.datadoghq.com/agent/configuration/secrets-management/?tab=linux#script-for-reading-from-multiple-secret-providers |
```

File: docs/configuration.v2alpha1.md [hunk 3]
```diff
@@ -156,6 +156,11 @@ spec:
...
+| features.secretsBackend.enableGlobalPermissions | EnableGlobalPermissions defines whether to create a global permission allowing Datadog agents to read all Kubernetes secrets. Default: false |
```
Comment (neko-dd) [inline docs/configuration.v2alpha1.md:161]:
```suggestion
| features.secretsBackend.enableGlobalPermissions | Whether to create a global permission allowing Datadog agents to read all Kubernetes secrets. Default: `false` |
```

File: docs/configuration.v2alpha1.md [hunk 4]
```diff
@@ -156,6 +156,11 @@ spec:
...
+| features.secretsBackend.roles | Roles defines roles for Datadog to read the specified secrets, replacing `enableGlobalPermissions`. They are defined as a list of namespace/secrets. Each defined namespace needs to be present in the DatadogAgent controller using `WATCH_NAMESPACE` / `DD_AGENT_WATCH_NAMESPACE`. See also: https://github.com/DataDog/datadog-operator/blob/main/docs/secret_management.md#how-to-deploy-the-agent-components-using-the-secret-backend-feature-with-datadogagent |
```
Comment (neko-dd) [inline docs/configuration.v2alpha1.md:162]:
```suggestion
| features.secretsBackend.roles | Roles for Datadog to read the specified secrets, replacing `enableGlobalPermissions`. They are defined as a list of `namespace/secrets`. Each defined namespace needs to be present in the DatadogAgent controller using `WATCH_NAMESPACE` or `DD_AGENT_WATCH_NAMESPACE`. See also: https://github.com/DataDog/datadog-operator/blob/main/docs/secret_management.md#how-to-deploy-the-agent-components-using-the-secret-backend-feature-with-datadogagent |
```

File: docs/configuration.v2alpha1.md [hunk 5]
```diff
@@ -156,6 +156,11 @@ spec:
...
+| features.secretsBackend.timeout | Timeout defines the command timeout in seconds. Default: 30 |
```
Comment (neko-dd) [inline docs/configuration.v2alpha1.md:163]:
```suggestion
| features.secretsBackend.timeout | The command timeout in seconds. Default: `30` |
```

File: docs/configuration.v2alpha1.md (hunk 1 above)
Comment (tbavelier) [reply docs/configuration.v2alpha1.md]: Fixed by [`ef90cdf` (#1333)](https://github.com/DataDog/datadog-operator/pull/1333/commits/ef90cdf9200d7a1f5a5c8c99a365f8de90226fad)

File: docs/configuration.v2alpha1.md (hunk 2 above)
Comment (tbavelier) [reply docs/configuration.v2alpha1.md]: Fixed by [`ef90cdf` (#1333)](https://github.com/DataDog/datadog-operator/pull/1333/commits/ef90cdf9200d7a1f5a5c8c99a365f8de90226fad)

File: docs/configuration.v2alpha1.md (hunk 3 above)
Comment (tbavelier) [reply docs/configuration.v2alpha1.md]: Fixed by [`ef90cdf` (#1333)](https://github.com/DataDog/datadog-operator/pull/1333/commits/ef90cdf9200d7a1f5a5c8c99a365f8de90226fad)

File: docs/configuration.v2alpha1.md (hunk 4 above)
Comment (tbavelier) [reply docs/configuration.v2alpha1.md]: Fixed by [`ef90cdf` (#1333)](https://github.com/DataDog/datadog-operator/pull/1333/commits/ef90cdf9200d7a1f5a5c8c99a365f8de90226fad)

File: docs/configuration.v2alpha1.md (hunk 5 above)
Comment (tbavelier) [reply docs/configuration.v2alpha1.md]: Fixed by [`ef90cdf` (#1333)](https://github.com/DataDog/datadog-operator/pull/1333/commits/ef90cdf9200d7a1f5a5c8c99a365f8de90226fad)

**Review by fanny-jiang — COMMENTED** [review]
Looks great! I only have one comment about the config naming

File: apis/datadoghq/v2alpha1/datadogagent_types.go [hunk 6]
```diff
@@ -100,6 +100,8 @@ type DatadogFeatures struct {
 	PrometheusScrape *PrometheusScrapeFeatureConfig `json:"prometheusScrape,omitempty"`
 	// HelmCheck configuration.
 	HelmCheck *HelmCheckFeatureConfig `json:"helmCheck,omitempty"`
+	// SecretsBackend configuration.
+	SecretsBackend *SecretsBackendFeatureConfig `json:"secretsBackend,omitempty"`
```
[the remark this answers is not in the document]
Comment (tbavelier) [reply apis/datadoghq/v2alpha1/datadogagent_types.go]: Done in https://github.com/DataDog/datadog-operator/pull/1333/commits/b2bfc5aa022fd3a26ff83d03e0a7d81f9f2b256f

**Review by fanny-jiang — CHANGES_REQUESTED** [review]
Also, can we move `secretBackend` from `features` to `global`?

**Comment by tbavelier** [comment]
Closing in favor of https://github.com/DataDog/datadog-operator/pull/1395 : moving from a feature to `spec.global` per comment

**Pull request closed without merging**